locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -2862,15 +2863,61 @@ protected String getFirstReachableAddress(NodeMetadata node, ConfigBag setup) {
-            result = JcloudsUtil.getFirstReachableAddress(node, timeout);
-            LOG.debug("Using first-reachable address "+result+" for node "+node+" in "+this);
+
+            Predicate<HostAndPort> pollForFirstReachableHostAndPortPredicate;
+            if (setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE) != null) {
+                LOG.debug("Using pollForFirstReachableAddress.predicate supplied from config for location " + this + " "
+                        + POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + " : " + setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE));
+                pollForFirstReachableHostAndPortPredicate = setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE);
+            } else {
+                LOG.debug("Using pollForFirstReachableAddress.predicate.type supplied from config for location " + this + " " + POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE.getName()
+                        + " : " + setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE));
+
+                List<Constructor<?>> constructors = Arrays.asList(setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE).getConstructors());
+                Optional<Constructor<?>> emptyConstructor= Iterables.tryFind(constructors, new Predicate<Constructor<?>>() {
+                    @Override public boolean apply(Constructor<?> constructor) {
+                        return constructor.getParameterTypes().length == 0;
+                    }
+                });
+                Optional<Constructor<?>> mapConstructor = Iterables.tryFind(constructors, new Predicate<Constructor<?>>() {
+                    @Override public boolean apply(Constructor<?> constructor) {
+                        if (constructor.getParameterTypes().length == 1) {
+                            return constructor.getParameterTypes()[0].isAssignableFrom(Map.class);
+                        } else {
+                            return false;
+                        }
+                    }
+                });
+                if (mapConstructor.isPresent()) {
+                    Map<String, Object> args = MutableMap.of();
+                    ConfigUtils.addUnprefixedConfigKeyInConfigBack(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + ".", setup, args);
+                    try {
+                        pollForFirstReachableHostAndPortPredicate = setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE).getConstructor(Map.class).newInstance(args);
+                    } catch (ReflectiveOperationException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                } else {
+                    try {
+                        pollForFirstReachableHostAndPortPredicate = (Predicate<HostAndPort>)emptyConstructor.get().newInstance();
+                    } catch (ReflectiveOperationException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                }
```

Comment:
Could log a user-friendly message for the case where both map and empty constructors are missing.